swarm/metrics.py

```diff
@@ -0,0 +1,165 @@
+            In use case 2 this should be a numpy array of losses, which can also be yeilded by core.swarm_train
+            It will have two dimensions, of structure [bee, epoch].
+            In use case 3 this should be a numpy array of summary statistics across bees at all points in domain x.
+            It will have two dimensions, of structure [epoch/batch, x_step]
+        func_string: A string indicating the type of summary statistic desired.
+            Currently allowable includes 'min', 'max', 'mean', 'median', 'std', and 'range'
+        across: Takes ony 'Indicate whether you'd like to summarise across the bees (as in use case 1 and 2),
+            or summarise across the a value for each point in domain x.
+
+    Returns: a numpy array of two dimensions, of structure [epoch/batch, x_step] that gives the
+        summary evaluation for each point in x space for all the bees at that epoch of training.
+
+    """
+    valid_axes = {'bees', 'x'}
+    if across not in valid_axes:
+        raise ValueError("argument `across` must be one of %r." % valid_axes)
+
+    # Create a simple dict of likely functions
+    funcs = {
+        "mean": np.mean,
+        "std": np.std,
+        "median": np.median,
+        "min": np.min,
+        "max": np.max
+    }
+
+    # Identify the suitable axis to summarise across
+    if across == "bees":
+        axis_choice = 0
+    elif across == "x":
+        axis_choice = len(data.shape) - 1
+
+    # Create summary of data
+    if func_string in funcs.keys():
+        return np.apply_along_axis(funcs[func_string], axis_choice, data)
+    elif func_string == 'range':
+        point_maxs = np.apply_along_axis(np.max, axis_choice, data)
+        point_mins = np.apply_along_axis(np.min, axis_choice, data)
+        return point_maxs - point_mins
+    else:
+        raise ValueError("Function not in currently allowable of min', 'max', 'mean', 'median', 'std', and 'range'")
+
+
+def rms_error(swarm_ypreds: np.ndarray, y: np.ndarray) -> np.ndarray:
+    """
+    This is a function that calculates the root-mean sqared error for a prediction array, given true y.
+    It averages along the all the point predictions in x for a set of predictions.
+    Args:
+        swarm_ypreds: A numpy array of predictions. This can have different numbers of dimensions,
+            eg [bee, epoch, x_step] or [epoch, x_step], so long as the final dimension represents the
+            predictions at different x values, and has same length as y.
+        y: The true values predictions. A 1-d array.
+
+    Returns: A numpy array of dimension one less than the input `swarm_ypreds`, with the evaluated root-mean-squared
+    loss across all predictions.
+    """
+    dims = len(swarm_ypreds.shape)
+    squares = np.square(np.subtract(swarm_ypreds, y))
+    losses = np.apply_along_axis(np.mean, dims-1, squares)
+    return losses
+
+
+def iteration_finder(epoch_scores: [np.ndarray, list],
```

Comment:
You want `Union[np.ndarray,list]` here I think - your annotation indicates a (list of array, list) which doesn't seem right. I also think you can just ignore the list being a type and assume `np.array` for everything

Reply:
Ah ok, yes I think this is helpful. I did look at a bunch of type-hinting examples. For some reason didn't find this Union notation. Glad you understand what I was trying to do. I certainly mean 'either or'. If it's safe to assume that everything is going to be strictly numpy, I guess then it could simplify things. I did consider that. A small part of inclination to include lists came from my noticing that for most small vectors (less than thousands), base python and lists is actually noticeably more performant than numpy. So perhaps the option is ok to allow for, if little extra cost? (And I can hint it correctly ;-)

Reply:
Pycharm doesn't seem to recognise Union... does it require a package? I might just go for numpy alone.

Reply:
Found the import requirement below! no probs